Publicly Verifiable UCAN Timestamps.md

```diff
@@ -0,0 +1,39 @@
+# Publicly verifiable timestamps for UCANs with Drand
+
+![](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)
+
+## Editors
+
+- [Alan Shaw](https://github.com/alanshaw), [DAG House](https://dag.house/)
+
+## Authors
+
+- [Alan Shaw](https://github.com/alanshaw), [DAG House](https://dag.house/)
+
+# Abstract
+
+Being able to verify that a UCAN invocation happened within a time window gives us some assurances.
+
+Enabling public verification that a UCAN was created on or after a specific time also allows services to be open and transparent.
+
+[Drand](https://drand.love/) is a distributed randomness beacon. Drand nodes produce collective, publicly verifiable, unbiased, unpredictable random values at fixed intervals using bilinear pairings and threshold cryptography.
+
+By including a drand randomness round in a signed UCAN, anyone can prove that it was created _on_ or _after_ that randomness round was generated. There’s a number of uses for this, an example is allowing the web3.storage service to account for `store/add` invocations, publicly proving that we were asked to store data within a given time frame.
+
+Services may choose to reject UCANs with timestamps that do not correspond to a recent randomness round. In this case it's reasonable to assume that the UCAN was _probably_ not generated long after the included randomness round. There is no incentive to do so if it is likely to be rejected on receipt.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+1. Issuers **must** include a drand randomness round in UCAN [`nnc`](https://github.com/ucan-wg/spec#323-nonce) field.
+
+1. Issuers **should** include the drand randomness _round number_ in the [`nb`](https://github.com/ucan-wg/spec#241-nb-non-normative-fields) field.
```

Comment:
Is this intentionally vague about *how* to include this in `nb`? i.e. under what key (e.g. `nb.drand` or `nb.time`)

Reply:
What's the motivation of this **should** vs not including it at all? If we can agree on a place to store it in `nnc` or `fct`, do we really need to strongly encourage it in `nb` too?

Reply:
@gobengo the proposal is to put the actual random _value_ in `nnc` and the round _number_ in `nb`. Not intentional under which key in `nb` -  needs specifying (I just forgot).

If `fct` is more appropriate place for any/all of this then we should put it there! I think it probably a better place seeing as this is not really a nonce but a timestamp.